File: shared/nav.py
"""Shared nav helpers, imported by every feature's build.py.
Not a script: no CLI, imported like a plain module.

esc() is used by every feature for HTML-escaping. require() is used by
every feature to validate a required field on a hand-edited JSON record
with a friendly message (record + missing field) instead of a raw
KeyError traceback. render_nav() builds the one shared nav bar shape
every page uses — see its own docstring and requirements/public.md ->
Navigation.
"""
import json
from datetime import date

from trip import TRIP_START, QUARTERS, QUARTER_NAMES, quarter_screen_id, format_date_jump, format_time_range


def esc(s):
    return str(s).replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;').replace('"', '&quot;')


def require(record, field, label):
    """Return record[field], or raise a friendly ValueError naming `label`
    (a human description of which record, e.g. "Person at index 3 (id 7)
    in shared/data/people.json") instead of letting a bare KeyError from
    record[field] surface as a raw traceback during hand-edited data entry."""
    if field not in record:
        raise ValueError(f"{label} is missing required field {field!r}.")
    return record[field]


def join_names(names):
    """Plain-English join for a short list of names — "Ann", "Ann & Bo", or
    "Ann, Bo & Cy". Used by both timeline/scripts/build.py (grouping
    several people who share one arrival/departure leg into a single line
    — see render_travel_row()) and attendees/scripts/build.py (an airport
    run's passenger list — see airport_run_line()), so the two can't drift
    on the join style. Not esc()'d here — callers pass already-escaped
    names."""
    if len(names) == 1:
        return names[0]
    return f"{', '.join(names[:-1])} & {names[-1]}"


def render_folks_menu(people, travel, *, timeline_prefix, attendees_prefix):
    """The Folks split control's panel content — one shape, one roster,
    used identically on every page (Home/Timeline, Family Tree, Details)
    now, replacing what used to be two different implementations that had
    to be kept in sync by hand (a travel.json-sourced list on Home/
    Timeline, a people.json-sourced one everywhere else) — see
    requirements/public.md -> Navigation -> Folks panel.

    Every ATTENDING person in shared/data/people.json, no one left out for
    lacking a travel.json entry yet. Each entry is a plain-text name plus
    two links: "Timeline" jumps to their arrival's quarter screen, or the
    trip's very first quarter screen if they have no arrival entry at all
    (same fallback the old Home/Timeline variant already used for this
    case); "Detail" goes to their own Attendees page. No status text
    (no "Collecting facts") — that distinction now lives only on the
    Family Tree box itself and the person's own Attendees page, not
    narrated a second time here.

    `timeline_prefix` is the same value the caller already passes to
    render_nav() below ("" on site/index.html itself, e.g. "../index.html"
    elsewhere) — a same-page anchor or a cross-page link to the exact same
    target either way. Display order is NOT decided here: every entry
    renders in whatever order `people` is passed in, and shared/nav.js
    shuffles the actual DOM order fresh on every page load (see
    requirements/public.md -> Navigation -> Folks panel) — a random order
    baked in at build time would be the same every time this static page
    is loaded, which isn't actually random."""
    travel_by_person = {t["person_id"]: t for t in travel}
    attending = [p for p in people if p.get("attending")]
    if not attending:
        return ""

    entries = []
    for person in attending:
        entry = travel_by_person.get(person["id"])
        arrival = entry.get("arrival") if entry else None
        if arrival:
            target_date, target_quarter = arrival["date"], arrival["quarter"]
        else:
            target_date, target_quarter = TRIP_START.isoformat(), QUARTERS[0]
        timeline_href = f"{timeline_prefix}#{quarter_screen_id(target_date, target_quarter)}"
        entries.append(
            f'<span class="jump-person">'
            f'<span class="jump-person-name">{esc(person["name"])}</span>'
            f'<span class="jump-person-actions">'
            f'<a href="{timeline_href}">Timeline</a>'
            f'<a href="{attendees_prefix}{person["id"]}.html">Detail</a>'
            f'</span>'
            f'</span>'
        )
    return "".join(entries)


def render_milestones_menu(people, travel, *, timeline_prefix, attendees_prefix):
    """The Milestones panel's two lists — every attending person's own
    arrival and every attending person's own departure, one flat
    chronological row per person, oldest first (see
    requirements/public.md -> Navigation -> Milestones panel). Computed
    once here and reused identically by every feature's build.py, same
    reasoning render_folks_menu() above already documents (one shape, one
    roster, no per-feature copy to drift out of sync).

    An excursion's own `return` leg counts as an arrival and its `depart`
    leg counts as a departure, same as everywhere else on the site (the
    Timeline's own Arriving:/Departing: rows, driving_assignments() on the
    Attendees page) — someone with a mid-stay round trip shows up twice in
    these two lists. Not-attending people, and anyone with no arrival (or
    no departure) leg at all, simply don't appear in that one list — no
    synthesized fallback the way the Folks panel's random-click target
    needs one.

    Deliberately ONE ROW PER PERSON, never grouped the way the Timeline's
    own rows combine several people sharing one real trip (see
    requirements/public.md -> Home & Timeline -> Row-by-row rules ->
    Arrivals) — this panel is a scannable flat roster of individual facts,
    not a day-quarter canvas, so there's no "same trip" concept to fold
    rows into here.

    Returns (arrivals_html, departures_html) — render_nav() below wraps
    each in its own tab panel."""
    travel_by_person = {t["person_id"]: t for t in travel}
    attending = [p for p in people if p.get("attending")]

    def sort_key(item):
        _, leg = item
        return (leg["date"], QUARTERS.index(leg["quarter"]), leg.get("time_range") or [""])

    def render_list(events):
        events.sort(key=sort_key)
        rows = []
        for person, leg in events:
            d = date.fromisoformat(leg["date"])
            time_range = leg.get("time_range")
            if time_range:
                when = f"{esc(format_date_jump(d))} · {esc(format_time_range(time_range))}"
            else:
                qname = QUARTER_NAMES[leg["quarter"]]
                when = esc(format_date_jump(d)) + (f" · {esc(qname)}" if qname else "")
            timeline_href = f"{timeline_prefix}#{quarter_screen_id(leg['date'], leg['quarter'])}"
            rows.append(
                f'<span class="jump-person milestone-row">'
                f'<span class="jump-person-name">{esc(person["name"])}</span>'
                f'<span class="milestone-time">{when}</span>'
                f'<span class="jump-person-actions">'
                f'<a href="{timeline_href}">Timeline</a>'
                f'<a href="{attendees_prefix}{person["id"]}.html">Detail</a>'
                f'</span>'
                f'</span>'
            )
        return "".join(rows)

    arrivals, departures = [], []
    for person in attending:
        entry = travel_by_person.get(person["id"])
        if not entry:
            continue
        if entry.get("arrival"):
            arrivals.append((person, entry["arrival"]))
        if entry.get("departure"):
            departures.append((person, entry["departure"]))
        for exc in entry.get("excursions", []):
            departures.append((person, exc["depart"]))
            arrivals.append((person, exc["return"]))

    return render_list(arrivals), render_list(departures)


def render_nav(*, mc26_href, timeline_prefix, tree_href, trip_start, trip_end,
                jump_panel_html, folks_panel_html, milestones_panel_html, attending_people, attendees_prefix,
                include_play):
    """The site's one shared nav bar shape — MC26 (fixed identity + link to
    Home), Timeline (split control: label jumps to "now", caret opens a
    panel with the day/quarter jump list and, on site/index.html only,
    play/pause), Folks (split control: label jumps to a random attending
    person, caret opens a person-picker panel), Milestones (caret-only
    disclosure — no default click of its own, see render_milestones_menu()
    above — opens a panel with two tab-switched flat lists, every
    arrival/every departure), Tree (plain link) — see
    requirements/public.md -> Navigation. Identical shape on every page;
    only hrefs/prefixes and panel contents differ, supplied by the caller,
    so the three build scripts can never drift apart on what this looks
    like.

    `timeline_prefix` — relative path to site/index.html: "" when this
    page IS index.html (shared/nav.js reads the empty data-prefix and
    scrolls in place instead of navigating), otherwise e.g. "../index.html".
    `jump_panel_html` / `folks_panel_html` — pre-rendered panel contents:
    the day/quarter jump list is still feature-specific (see
    timeline/scripts/build.py's render_jump_panel()), but the Folks panel
    is now this same module's own render_folks_menu() above, called
    identically by every feature. `milestones_panel_html` is a
    (arrivals_html, departures_html) pair — this module's own
    render_milestones_menu() above, same "computed once, called
    identically by every feature" shape as the Folks panel.
    `attending_people` — the full attending roster as a list of dicts with
    at least 'id'/'name', for the Folks split control's random-click (the
    same roster regardless of which Folks panel variant is shown), and to
    decide whether the Milestones item renders at all (no roster, nothing
    to list).
    `include_play` — True only for site/index.html, the one page with an
    actual scroll sequence to auto-advance through.
    """
    play_html = (
        '<button type="button" id="run-toggle" class="run-toggle" aria-label="Play">▶</button>'
        if include_play else ""
    )
    timeline_split = f"""<span class="nav-split current-quarter-menu">
<button type="button" id="timeline-jump" class="nav-split-label current-quarter-label" data-trip-start="{esc(trip_start)}" data-trip-end="{esc(trip_end)}" data-prefix="{esc(timeline_prefix)}"><span class="cq-date" id="cq-date"></span><span class="cq-day" id="cq-day">Timeline</span></button>
<details class="jump-menu">
<summary><span class="nav-caret" id="timeline-caret">▾</span></summary>
<div class="jump-panel">{jump_panel_html}{play_html}</div>
</details>
</span>"""

    folks_split = ""
    if attending_people:
        people_attr = esc(json.dumps([{"id": p["id"], "name": p["name"]} for p in attending_people]))
        folks_split = f"""<span class="nav-split">
<button type="button" id="folks-random" class="nav-split-label" data-attendees-prefix="{esc(attendees_prefix)}" data-people="{people_attr}">Folks</button>
<details class="jump-menu">
<summary><span class="nav-caret">▾</span></summary>
<div class="jump-panel"><div class="folks-list">{folks_panel_html}</div></div>
</details>
</span>"""

    milestones_menu = ""
    if attending_people:
        arrivals_html, departures_html = milestones_panel_html
        milestones_menu = f"""<details class="jump-menu">
<summary><span class="nav-split-label">Milestones</span> <span class="nav-caret">▾</span></summary>
<div class="jump-panel milestones-panel">
<div class="milestones-tabs">
<button type="button" class="milestones-tab is-active" data-milestones-tab="arrivals">Arrivals</button>
<button type="button" class="milestones-tab" data-milestones-tab="departures">Departures</button>
</div>
<div class="milestones-list" data-milestones-panel="arrivals">{arrivals_html}</div>
<div class="milestones-list" data-milestones-panel="departures" hidden>{departures_html}</div>
</div>
</details>"""

    return f"""<nav class="site-nav">
<a href="{mc26_href}" class="site-title">MC26</a>
{timeline_split}
{folks_split}
{milestones_menu}
<a href="{tree_href}">Tree</a>
</nav>"""

File: shared/nav.js
// Shared client-side nav logic — inlined into every page at build time
// (see technical.md -> Where CSS and JS changes go). currentQuarterSectionId()
// is a plain top-level function (not wrapped in DOMContentLoaded), so it's
// available immediately once this script tag runs, before timeline/shared.js
// (site/index.html only, loaded after this file) needs to reuse it for its
// own Play/pause and transition-screen wiring — one computation, not two
// copies that could drift on "what quarter is it right now at the cottage."
//
// The DOMContentLoaded block below wires the "Timeline" and "Folks" split
// controls' own default-click behavior, and the shared close-a-panel
// behavior every .jump-menu disclosure needs (see requirements/public.md
// -> Navigation -> Closing a panel) — every page's nav bar has these,
// including Family Tree and Attendees pages, which have no other JS of
// their own to hang this off.

var NAV_TRIP_TIMEZONE = 'America/Moncton';

function navQuarterForHour(hour) {
  if (hour < 6) return '00-06';
  if (hour < 12) return '06-12';
  if (hour < 18) return '12-18';
  return '18-24';
}

// Intl.DateTimeFormat with an explicit timeZone reads the wall-clock
// date/hour AT THE COTTAGE regardless of the visitor's own system timezone
// — hourCycle: 'h23' pins the hour to a plain 0-23 range (some engines
// otherwise return "24" for midnight with hour12: false alone).
function navTripLocalParts(d) {
  var parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: NAV_TRIP_TIMEZONE,
    year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', hourCycle: 'h23'
  }).formatToParts(d);
  var map = {};
  parts.forEach(function (p) { map[p.type] = p.value; });
  return { iso: map.year + '-' + map.month + '-' + map.day, hour: parseInt(map.hour, 10) };
}

// tripStart/tripEnd are ISO date strings ("2026-08-01") read off whichever
// element triggered the call (its own data-trip-start/data-trip-end), not
// a single shared global — this same function works identically whether
// called from site/index.html or a page two directory levels away from it.
function currentQuarterSectionId(tripStart, tripEnd) {
  var now = navTripLocalParts(new Date());
  var iso = now.iso;
  var quarter = navQuarterForHour(now.hour);
  // Clamp into the trip window: before it starts snaps to the very first
  // quarter; after it ends snaps to the very last.
  if (iso < tripStart) {
    iso = tripStart;
    quarter = '00-06';
  } else if (iso > tripEnd) {
    iso = tripEnd;
    quarter = '18-24';
  }
  return 'qc-' + iso + '-' + quarter;
}

// A deliberate "jump to this section" action — used by every call site
// that can land more than one quarter screen away from where the visitor
// currently is: this file's own "Timeline" nav item click, and (in
// timeline/shared.js) the transition screen's two choices and every
// jump-menu link. Plain top-level function, same reasoning as
// currentQuarterSectionId() above — needed before timeline/shared.js's
// own DOMContentLoaded block runs, and shared rather than duplicated
// since call sites live in both files.
//
// Exists specifically to fix a real flashing-content bug (see
// requirements/public.md -> Home & Timeline -> Time-of-day background):
// timeline/shared.js's live scroll-position color blend repaints on
// every scroll event, which is exactly right for an organic scroll or a
// one-screen jump (a gentle cross-fade into the next quarter's color),
// but a browser's smooth-scroll animation takes roughly the same short
// duration to cross ANY distance — so a jump of, say, five screens fires
// that same per-frame repaint five times faster, racing through five
// quarters' colors in under a second. That reads as a strobe, not a
// cross-fade. window.__scrollJumpActive is the flag
// updateQuarterBlend() (timeline/shared.js) checks to skip its own
// per-frame work while a jump is in flight, so a multi-screen jump shows
// no color changes at all for its duration and settles in one clean
// step on the destination's own correct color the instant it lands —
// nothing skipped for a single-screen jump or organic scrolling, since
// isMultiScreenJump() below only sets the flag for a jump that actually
// crosses more than one quarter screen.
//
// Also suspends scroll-snap for the duration, same fix and same reason
// timeline/shared.js's play/pause run already applies (see its own
// startRun()/stopRun() comments and timeline/shared.css ->
// html.timeline-page.auto-scrolling): scrollIntoView({behavior:
// 'smooth'}) while an ancestor has scroll-snap-type: y mandatory active
// is a known bad combination on mobile Safari/iOS specifically — for a
// jump of more than one screen (the Home screen's "Aug 1st"/"Now"
// buttons, or any jump-menu link landing more than a screen away) it can
// silently fail to scroll at all, not just animate roughly, since the
// play/pause fix only ever covered its OWN repeated calls, not this
// function every other jump path (transition screen, jump-menu links,
// the Timeline nav item) actually goes through. Harmless on any page
// other than site/index.html — .auto-scrolling only matches CSS scoped
// under html.timeline-page (timeline/shared.css), which no other page's
// <html> carries. Unlike the blend suspension above, this part is NOT
// scoped to multi-screen jumps only — a single-screen jump goes through
// this exact same scrollIntoView(smooth) call and hits the identical
// Safari/iOS snap bug, so scroll-snap is suspended for every jump
// regardless of distance.
//
// isMultiScreenJump() reads target's own current bounding rect rather
// than tracking "which screen is current" as a separate piece of state:
// quarter screens stack at one viewport height each (100svh — see
// timeline/shared.css), so before any scrolling starts, a target at most
// one screen away always sits within about one viewport height of the
// current scroll position (roughly 0 for the same screen, roughly
// +/-100vh for an adjacent one); anything further is unambiguously a
// multi-screen jump. The 1.5x margin absorbs the small, expected slack
// from the nav bar's own height and mobile toolbar collapse/expand,
// without ever mistaking a genuinely-adjacent screen for a far one.
function isMultiScreenJump(target) {
  var top = target.getBoundingClientRect().top;
  return Math.abs(top) > window.innerHeight * 1.5;
}
var scrollJumpSettleTimer = null;
function jumpToSection(target) {
  if (!target) return;
  var suspendBlend = isMultiScreenJump(target);
  if (suspendBlend) {
    window.__scrollJumpActive = true;
  }
  document.documentElement.classList.add('auto-scrolling');
  if (scrollJumpSettleTimer) clearTimeout(scrollJumpSettleTimer);
  var settle = function () {
    if (suspendBlend) {
      window.__scrollJumpActive = false;
    }
    document.documentElement.classList.remove('auto-scrolling');
    document.removeEventListener('scrollend', settle);
    // A one-off, definitive correction for the Structures stage (and the
    // nav label alongside it) — timeline/shared.js exposes this only on
    // site/index.html (the one page with quarter screens/a stage to
    // correct), so it's guarded here rather than assumed to exist; see
    // that file's own window.__syncStructuresStage comment for why this
    // can't just piggyback on the synthetic 'scroll' event below the way
    // the blend does — IntersectionObserver doesn't listen for 'scroll'
    // at all, so nudging one wouldn't reach it.
    if (typeof window.__syncStructuresStage === 'function') {
      window.__syncStructuresStage();
    }
    // Nudge timeline/shared.js's own scroll listener to recompute now
    // that we've landed — without this, a suspended blend would stay
    // frozen at whatever it was showing right before the jump until the
    // visitor's next real scroll/swipe. Harmless to dispatch even when
    // suspendBlend was false (the blend was never frozen in the first
    // place, so this just repaints with the same answer it already had).
    window.dispatchEvent(new Event('scroll'));
  };
  // 'scrollend' (supported in every current evergreen browser) fires the
  // instant the browser's own smooth-scroll animation actually finishes
  // — far more reliable than guessing a fixed duration. The setTimeout
  // is only a fallback for a browser missing it entirely.
  if ('onscrollend' in window) {
    document.addEventListener('scrollend', settle, { once: true });
  } else {
    scrollJumpSettleTimer = setTimeout(settle, 700);
  }
  target.scrollIntoView({ behavior: 'smooth', block: 'start' });
}

document.addEventListener('DOMContentLoaded', function () {
  // "Timeline" nav item's own click (see requirements/public.md ->
  // Navigation) — jumps to "now": scrolls in place when already on
  // site/index.html (data-prefix empty), or navigates there when on
  // Family Tree/a Details page (data-prefix e.g. "../index.html").
  var timelineJump = document.getElementById('timeline-jump');
  if (timelineJump) {
    timelineJump.addEventListener('click', function () {
      var id = currentQuarterSectionId(timelineJump.dataset.tripStart, timelineJump.dataset.tripEnd);
      var prefix = timelineJump.dataset.prefix || '';
      if (prefix) {
        window.location.href = prefix + '#' + id;
      } else {
        jumpToSection(document.getElementById(id));
      }
    });
  }

  // "Folks" nav item's own click — jumps to a random attending person's
  // Details page, computed fresh each time. data-people is the full
  // attending roster ([{id, name}, ...], JSON-encoded); data-attendees-
  // prefix the page-relative path to site/attendees/.
  var folksRandom = document.getElementById('folks-random');
  if (folksRandom) {
    folksRandom.addEventListener('click', function () {
      var people = JSON.parse(folksRandom.dataset.people || '[]');
      if (!people.length) return;
      var person = people[Math.floor(Math.random() * people.length)];
      window.location.href = (folksRandom.dataset.attendeesPrefix || '') + person.id + '.html';
    });
  }

  // Folks panel — freshly shuffled display order on every page load (see
  // requirements/public.md -> Navigation -> Folks panel), so no single
  // person is always first or last. A build-time order would be the same
  // every time this static page loads, which isn't actually random — has
  // to happen here instead. Fisher-Yates on the real DOM nodes, each
  // person shuffled independently (not grouped by couple/family), then
  // re-appended in the new order; re-appending an already-attached node
  // moves it rather than duplicating it.
  var folksList = document.querySelector('.folks-list');
  if (folksList) {
    var people = Array.prototype.slice.call(folksList.children);
    for (var i = people.length - 1; i > 0; i--) {
      var j = Math.floor(Math.random() * (i + 1));
      var tmp = people[i];
      people[i] = people[j];
      people[j] = tmp;
    }
    people.forEach(function (el) { folksList.appendChild(el); });
  }

  // Closing a disclosure (see requirements/public.md -> Navigation ->
  // Closing a panel) — a plain <details> does neither of these two on
  // its own: opening one leaves any other already-open one (a nav panel,
  // or — see below — a person-chip's revealed name) sitting open too,
  // overlapping on screen, and clicking anywhere outside an open one
  // does nothing at all — the only way to close it is clicking the exact
  // same tiny trigger again. Both read as "hard to close." Lives here
  // (not timeline/shared.js) so it applies on every page with one of
  // these, including Family Tree and Attendees pages, which have no
  // other JS of their own.
  //
  // Covers two visually different but mechanically identical things:
  // the nav's own split-control panels (.jump-menu — Timeline's day/
  // quarter jump list, Folks, Milestones) and a person-chip's tap-to-
  // reveal name (.person-chip — render_person_chip() in
  // timeline/scripts/build.py, see requirements/public.md -> The
  // Structures stage -> Nested box display and -> Row-by-row rules ->
  // Arrivals) — both are just <details>/<summary>, so one shared "only
  // one open at a time, click away or Escape closes it" rule covers
  // every one of them site-wide, not a second bespoke mechanism for
  // chips. A page can have a couple hundred person-chips (every occupant
  // ever shown on the Structures stage, every traveler on every Arrivals/
  // Departures row, all pre-rendered — see technical.md -> The
  // Structures stage) — trivial to attach one listener each on a static
  // page this size.
  var disclosures = document.querySelectorAll('.jump-menu, .person-chip');
  if (disclosures.length) {
    // Opening one closes any other open one first — 'toggle' fires
    // whenever a <details>'s own open state changes, native or scripted,
    // so this also catches one opened programmatically (e.g. a future
    // change), not just a direct click on its own trigger.
    disclosures.forEach(function (menu) {
      menu.addEventListener('toggle', function () {
        if (!menu.open) return;
        disclosures.forEach(function (other) {
          if (other !== menu) other.open = false;
        });
      });
    });
    // A click anywhere outside every open one closes it — checked via
    // .contains() so a click on its own trigger or a link inside it is
    // correctly left alone (the trigger's native toggle, or
    // timeline/shared.js's own close-after-acting logic, handle those).
    document.addEventListener('click', function (event) {
      disclosures.forEach(function (menu) {
        if (menu.open && !menu.contains(event.target)) {
          menu.open = false;
        }
      });
    });
    // Escape closes whatever's open — the keyboard equivalent of
    // clicking away, for anyone not using touch/mouse.
    document.addEventListener('keydown', function (event) {
      if (event.key !== 'Escape') return;
      disclosures.forEach(function (menu) { menu.open = false; });
    });
  }
});
